staging/src/k8s.io/client-go/resolver/resolver.go

```diff
@@ -0,0 +1,349 @@
+	// that use a load balancer for bootstrapping
+	r.refreshCache(ctx)
+	utilnet.CloseIdleConnectionsFor(r.client.Client.Transport)
+	return resolver, nil
+}
+
+func setConfigDefaults(config *rest.Config) error {
+	gv := v1.SchemeGroupVersion
+	config.GroupVersion = &gv
+	config.APIPath = "/api"
+	config.NegotiatedSerializer = scheme.Codecs.WithoutConversion()
+
+	if config.UserAgent == "" {
+		config.UserAgent = rest.DefaultKubernetesUserAgent()
+	}
+
+	return nil
+}
+
+func (r *resolver) lookupIP(ctx context.Context, network, host string) ([]net.IP, error) {
+	// Use the default resolver if is not trying to resolve the configured API server hostname
+	if !strings.HasPrefix(host, r.host) {
+		klog.V(7).Infof("Resolver Trace: use default resolver for host %s, different than API server hostname %s", host, r.host)
+		return net.DefaultResolver.LookupIP(ctx, network, host)
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	// Use the default resolver if the cache is empty and try to renew the cache.
+	if len(r.cache) == 0 {
+		klog.V(7).Infof("Resolver Trace: use default resolver for host %s, cache is empty", host)
+		addrs, err := net.DefaultResolver.LookupIP(ctx, network, host)
+		if err != nil {
+			return addrs, err
+		}
+		// don't try to renew the cache until we know it's possible
+		if len(addrs) > 0 {
+			klog.V(7).Infof("Resolver Trace: refreshing cache for first time for host %s", host)
+			go r.refreshCache(ctx)
+		}
+		return addrs, nil
+	}
+
+	// Check if is the first time we query the cache and record the time, it is a fresh cache
+	// so don't try to refresh it again, if something goes wrong next time will be refreshed.
+	// If is not it the first time, it means that previous connection didn't work out
+	// so we return the IPs in a different order to reach a different API server.
+	if r.timestamp.IsZero() {
+		r.timestamp = time.Now()
+	} else {
+		rand.Shuffle(len(r.cache), func(i, j int) {
+			r.cache[i], r.cache[j] = r.cache[j], r.cache[i]
+		})
+		if r.refresh.get() {
+			klog.V(7).Infof("Resolver Trace: refreshing cache for host %s", host)
+			go r.refreshCache(ctx)
+		}
+	}
+
+	// Return the IP addresses from the cache.
+	ips := make([]net.IP, len(r.cache))
+	copy(ips, r.cache)
+	klog.V(7).Infof("Resolver Trace: host %s resolves to %v", host, ips)
+	return ips, nil
+}
+
+// refreshCache refresh the cache with the API server IP addresses.
+// If it can not refresh the IPs because of network errors during
+// a predefined time, it falls back to the default resolver.
+// If it can not refresh the IPs because of other type of errors, per
+// example, it is able to connect to the API server but this is not ready
+// or is not able to reply, it shuffles the IPs so it will retry randomly.
+// If there are no errors, local IP addresses are returned first, so it
+// favors direct connectivity.
+func (r *resolver) refreshCache(ctx context.Context) {
+	// avoid multiple refresh queries in paralell
+	r.refresh.set(false)
```

Comment:
this races. I would set it to false before the `go refreshCache` call.

Reply:
ack